File: C/Algorithm/Algorithm I/695. Max Area of Island.c
#define MAX_QUEUE_SIZE 2500

#define MAX(X, Y) (X > Y ? X : Y)

typedef struct {
    int row;
    int col;
} Pair;

typedef struct {
    Pair data[MAX_QUEUE_SIZE];
    int front;
    int rear;
} Queue;

Queue* createQueue() {
    Queue* queue = (Queue*) malloc (sizeof(Queue));
    queue->front = -1;
    queue->rear = -1;
    return queue;
}

void enqueue(Queue* queue, int row, int col) {
    if (queue->rear == MAX_QUEUE_SIZE - 1)
        return;

    if (queue->front == -1)
        queue->front = 0;

    queue->rear++;
    queue->data[queue->rear].row = row;
    queue->data[queue->rear].col = col;
}

Pair dequeue(Queue* queue) {
    Pair item;
    item.row = -1;
    item.col = -1;

    if (queue->front == -1 || queue->front > queue->rear)
        return item;

    item = queue->data[queue->front];
    queue->front++;
    return item;
}

int isValid(int row, int col, int gridSize, int* gridColSize) {
    return row >= 0 && row < gridSize && col >= 0 && col < gridColSize[row];
}

int maxAreaOfIsland(int** grid, int gridSize, int* gridColSize) {
    int maxArea = 0;
    int directions[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    for (int row = 0; row < gridSize; row++) {
        for (int column = 0; column < gridColSize[row]; column++) {
            if (grid[row][column] == 1) {
                int area = 0;
                Queue* queue = createQueue();
                enqueue(queue, row, column);
                
                grid[row][column] = 0;

                while (queue->front <= queue->rear) {
                    Pair current = dequeue(queue);
                    area++;

                    for (int k = 0; k < 4; k++) {
                        int newRow = current.row + directions[k][0];
                        int newCol = current.col + directions[k][1];

                        if (isValid(newRow, newCol, gridSize, gridColSize) && grid[newRow][newCol] == 1) {
                            enqueue(queue, newRow, newCol);
                            grid[newRow][newCol] = 0;
                        }
                    }
                }
                
                maxArea = MAX(maxArea, area);

                free(queue);
            }
        }
    }

    return maxArea;
}

File: Python/Algorithm/Algorithm I/695. Max Area of Island.py
import collections
from typing import List


class Solution:
    def isValidCell(self, row: int, column: int, ROWS: int, COLUMNS: int) -> bool:
        """
        Checks if the given cell coordinates are valid within the grid.

        Args:
            row: The row index of the cell.
            column: The column index of the cell.
            ROWS: The total number of rows in the grid.
            COLUMNS: The total number of columns in the grid.

        Returns:
            True if the cell is valid, False otherwise.

        Time Complexity: o(1) since we do 2 comparisons.
        Space Complexity: o(n) since we do not use any extra space.
        """
        isValidRow = 0 <= row < ROWS
        isValidColumn = 0 <= column < COLUMNS

        return isValidRow and isValidColumn

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        """
        Calculates the maximum area of an island in the binary matrix.

        Args:
            grid: The binary matrix representing land (1) and water (0).

        Returns:
            The maximum area of an island. If there is no island, returns 0.

        Time Complexity: o(m * n), where m is the number of rows and n is the number of columns in the grid.
        Space Complexity: O(m * n), where m is the number of rows and n is the number of columns in the grid.
        """
        ROWS = len(grid)
        COLUMNS = len(grid[0])
        DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        maxArea = 0

        for row in range(ROWS):
            for column in range(COLUMNS):
                if grid[row][column]:
                    grid[row][column] = 0

                    queue = collections.deque()
                    queue.append((row, column))
                    currentArea = 1

                    while queue and queue[0]:
                        row, column = queue.popleft()

                        for rowStep, columnStep in DIRECTIONS:
                            newRow = row + rowStep
                            newColumn = column + columnStep

                            if self.isValidCell(newRow, newColumn, ROWS, COLUMNS):
                                if grid[newRow][newColumn]:
                                    queue.append((newRow, newColumn))
                                    grid[newRow][newColumn] = 0
                                    currentArea += 1

                    maxArea = max(maxArea, currentArea)

        return maxArea
